### DISCUSSION TITLE
Add assumptions to a variable

### DISCUSSION THREAD

**Discussion by said-hadjout** [pyccel/pyccel#1805] [Ideas]
In Psydac we want to use negative indices without adding the `allow_negative_indices` decorator, we can introduce a mechanism to specify that a variable is intended to be positive. This will allow the compiler to optimize the code.
[Here](https://stackoverflow.com/questions/40447195/can-i-hint-the-optimizer-by-giving-the-range-of-an-integer) you can see an optimization done by the compiler by specifying the range of a variable.
Maybe we can provide that in Pyccel by doing something like
```python
x: 'int(0, inf)' = 5
```
This will indicate that `x` is positive

**Comment by EmilyBourne** [comment]
This sounds like a good idea, but I am not sure about the syntax. I think that achieving this with a type annotation is the right approach though

**Comment by EmilyBourne** [comment]
What constraints prevent you using `allow_negative_indices`?

**Reply by said-hadjout** [reply]
`allow_negative_indices` will slow down the code because the checks will be used in all arrays

**Reply by EmilyBourne** [reply]
The docs for `allow_negative_index` are here:
https://github.com/pyccel/pyccel/blob/devel/docs/decorators.md#allow-negative-index

It does not apply to the whole code, but only to the arrays specified in the argument. The example in the docs is:
```python
from pyccel.decorators import allow_negative_index
from numpy import array

@allow_negative_index('a')
def fun1(i : int, j : int):
    #////////negative indexing allowed////////
    a = array([1,2,3,4,5,6])
    print(a[i - j])
    #////////negative indexing disallowed. the generated code can cause a crash/compilation error.////////
    b = array([1,2,3,4,5,6])
    print(b[i - j])
```

**Reply by said-hadjout** [reply]
ah yes you are right

**Reply by yguclu** [reply]
Indeed the `@allow_negative_index` decorator applies to a single array. But in Psydac that array has 6 dimensions and it is updated inside 9 nested loops. Hence it is important to avoid any useless sign check.

**Reply by EmilyBourne** [reply]
As a temporary workaround maybe you can use a pointer to the slice that allows negative indexes?

**Reply by yguclu** [reply]
Not sure that helps. We would like to use the negative index (if possible) inside the nested loop 😅

**Reply by EmilyBourne** [reply]
That is what I had in mind. Maybe with an example we can understand each other? My suggested workaround is:
```python
@allow_negative_index('x_ptr')
def f(x : 'int[:,:,:]'):
    n,m,p = x.shape
    for i in range(n):
         for j in range(m):
              x_ptr = x[i,j]
              for k in range(p):
                        ...
```

**Reply by yguclu** [reply]
Your example reduces the number of `if` conditions which will be printed, but there is still one in the innermost loop where we access `x_ptr` with a negative index. Our current workaround is to calculate the actual positive index, but the Python code is less nice to read. Ultimately our request is not urgent at all, as we are really talking about syntactic sugar.

**Reply by EmilyBourne** [reply]
Can you share some code. I still don't understand what you are trying to achieve. It seems to me that you don't want to add an assumption to `x` but rather to the index that you are using?

**Reply by EmilyBourne** [reply]
Never mind, this is indeed what Said wrote

**Reply by yguclu** [reply]
Yes, the assumption is on the index itself (Said called it `x`, which is a bit confusing). We would like to declare an integer variable `pad` as positive, so we could treat
```python
a[pad : -pad]
```
as
```python
a[pad : a.shape[0] - pad]
```
without the need for an `if` condition. A more complicated example is
```python
a[pad + 1 : -(pad + 1)]
```
but this amounts to computing at **compile time** whether an expression is positive, negative, or undefined. Maybe we could do this just for array indexing, though.

**Comment by EmilyBourne** [comment]
Related discussion : https://github.com/pyccel/pyccel/discussions/1565

**Comment by said-hadjout** [comment]
One way to solve the issue of negative indices is to introduce the notion of valued variables.
if we have something like
```python
pad: 'Final' = 3
a[pad:-pad]
```
In the semantics stage, we should be able to know the value of the variable `pad` and in Fortran, it will be translated to a parameter

**Reply by EmilyBourne** [reply]
This should be easy to achieve and we already have the syntax. Any of the following are already valid.
```python
pad : Final[int] = 3
pad : 'Final[int]' = 3
pad : 'const int' = 3
```
They could be saved in a `pyccel.ast.variable.Constant`

**Reply by EmilyBourne** [reply]
This suggestion seems well defined enough to open an issue

**Reply by said-hadjout** [reply]
We need this in Psydac because we always hardcode some values to make the code faster, as a result, the code becomes unreadable sometimes. using this we should get the best of both worlds

**Comment by said-hadjout** [comment]
We can introduce a function called `assume` that takes a logical expression, we can use it in this manner
```python
def f(a:int[:], x:int):
    assume(x>0 and x <10)
   return a[-x]
``

**Reply by EmilyBourne** [reply]
Why not use `assert`?

**Reply by said-hadjout** [reply]
assert will raise an error if it is not True, I'm not sure we want to check that at runtime. This should work as a hint for the compiler to optimize the code at the compilation stage

**Reply by EmilyBourne** [reply]
It only raises an error in debug mode. By default we don't compile in debug mode. I'm not sure we add `-DNDEBUG` though. We should do that

**Reply by said-hadjout** [reply]
okay, sounds good then

**Reply by said-hadjout** [reply]
This idea is part of a proposal in C++ called [contracts](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2900r1.pdf), where the user can add assumptions to the code
